Fix export list

kierarad opened the pull request:
This is a fix for issue #5730 

In order to do this I've done 2 things: 
- I've added the capabilities to the plugin info for config repos so we can check if the plugin supports export
- I've filtered the dropdown on the pipeline config page to only show the plugins that support pipeline export

ketan commented on plugin-infra/go-plugin-access/src/main/java/com/thoughtworks/go/plugin/access/configrepo/ConfigRepoPluginInfoBuilder.java:
```diff
@@ -35,11 +36,14 @@ public ConfigRepoPluginInfoBuilder(ConfigRepoExtension extension) {
 
+    private Capabilities capabilities(String pluginId) {
+        return extension.getCapabilities(pluginId);
+    }
```
👍 

The intention behind this is to cache the calls made to the plugin, that always return the same data. Capabilities, get view, get icon, etc. I noticed a few days ago that `GoConfigPluginService` directly calls `getCapabilities` on the extension, instead of looking up the metadata store. https://github.com/gocd/gocd/blob/84ae8699545ee85320fbaeb0702acc4009d74032/server/src/main/java/com/thoughtworks/go/config/GoConfigPluginService.java#L62-L64

Would you mind changing that reference to use the metadata store instead (in a separate commit)?

kierarad replied:
Of course. Good catch

kierarad commented on server/dev/config/cipher.aes:
```diff
@@ -0,0 +1 @@
+e83d7343445b54ecdf4c5ca2afc31f47
```
@ibnc yes, good catch. I quickly pushed this up yesterday so that @ketan could see what I was referring to in regards to my versioning question. I definitely didn't mean to include this file. I should've noted that this PR is a WIP

kierarad commented:
@ketan I think this is ready for review (again). When I bumped the version I went ahead and removed v4 since it sounded like we are the only ones who use the API and we don't maintain any other previous versions of the plugin infos API. Let me know if you or @maheshp think this is a bad idea and I will add v4 back.

maheshp commented:
@kierarad I took a look at the PR, the changes around adding the capabilities to `ConfigRepoPluginInfo` seems fine and changes to use the cached capabilities in `GoConfigPluginService.java` is right.

Bumping up the PluginInfo needs to change. Sorry, I should have mentioned this earlier that we are moving away from rails, hence whenever we write a new API or bump up the version of an existing API we write it in Java/Spark. I am afraid we will have to do the same for PluginInfo as well.

Also, whenever we introduce a  new version of an existing API we deprecate and retain the current version for around 3 months and later remove it. Your changes actually removes the v4 of the endpoint and introduces v5 that would break for existing users.

For now I would suggest we split this PR into 2 parts, have one PR with changes to capabilities and get it merged and open a new PR for the PluginInfo V5 in Spark.

kierarad commented:
Thanks @maheshp! 

> Bumping up the PluginInfo needs to change. Sorry, I should have mentioned this earlier that we are moving away from rails, hence whenever we write a new API or bump up the version of an existing API we write it in Java/Spark. I am afraid we will have to do the same for PluginInfo as well.
>
Don't be sorry. This was my mistake. I should've known. That makes perfect sense and it's silly that I didn't consider that. 
 
> Also, whenever we introduce a new version of an existing API we deprecate and retain the current version for around 3 months and later remove it. Your changes actually removes the v4 of the endpoint and introduces v5 that would break for existing users.
> 
Ok this makes sense. I was unsure, so I should of asked. Again this is on me. Lesson learned, I was questioning my approach (obviously for good reason) so I should've asked instead of just doing. 

> For now I would suggest we split this PR into 2 parts, have one PR with changes to capabilities and get it merged and open a new PR for the PluginInfo V5 in Spark.
>
Will do. For now I've just added the V4 back to rails. I'll create a separate PR for the rewrite of the V5 in Spark.

maheshp commented:
Thanks @kierarad!

kierarad commented:
@maheshp is this one ok to merge?

arvindsv commented:
@kierarad Since you're moving this to Spark, why don't we just close this and just have this merged once it is moved to Spark? That way, the testing / reviewing effort of it can be reduced.

Since there are only 3 config repo plugins currently, and they all support the export feature, I think it's ok that it shows them all. Once the API in Spark is done, it'll be doing the right thing.

arvindsv commented:
Also: That way, you won't have to get rid of this, when moving to Spark.

kierarad commented:
@arvindsv There is still quite a bit of effort that will need to go into moving the PluginInfos API to Spark. Given the shift in priorities to improving the trial experience I don't foresee it will be completed for quite some time. Should go ahead and close this still?

arvindsv commented:
@kierarad Yes, I'd still do that. When someone gets a chance, they can pick it up again.

@maheshp We'll need to add this to the list and decide when to pick it back up.

ketan commented:
Can bring this back if needed.
